File: LeetCode/233. Number of Digit One/solution.py
class Solution:
    def countDigitOne(self, n: int) -> int:
        d, ans = 1, 0
        tmp = n//d
        while tmp > 0:
            x = tmp%10
            left = tmp//10
            right = n - (tmp*d)
            # print(left, right)
            ans += left*d
            if x == 1:
                ans += right+1
            elif x > 1:
                ans += d
            d *= 10
            tmp = n//d
        return ans

## find counts of 1 on digit x

File: LeetCode/233. Number of Digit One/solution.cpp
class Solution {
public:
    int countDigitOne(int n) {
        int cnt = 0;
        long long base = 1;
        while(base <= n){
            int prefix = n/base;
            int suffix = n%base;
            if (prefix%10 > 1){
                cnt += (prefix/10+1)*base;
            }
            else if(prefix%10 == 1){
                cnt += prefix/10*base;
                cnt += suffix+1;
            }
            else{
                cnt += prefix/10*base;
            }
            // cout << base << " " << cnt << endl;
            base *= 10;
        }
        return cnt;
    }
};